### PR TITLE
Make message formatter in category source optional + Increase MSI to 100%

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

File: tests/TranslatorTest.php [hunk 1]
```diff
@@ -559,6 +560,53 @@ public function testTranslationMissingMessage(): void
...
+        $translator = new Translator(
+            locale: 'en',
+            defaultMessageFormatter: new SimpleMessageFormatter()
+        );
+        $translator->addCategorySource($categorySourceWithFormatter);
+        $translator->addCategorySource($categorySourceWithoutFormatter);
+
+        $this->assertSame(
+            'Hello, {name}!',
+            $translator->translate('hello', ['name' => 'Kitty'], 'withFormatter')
```
Comment (darkdef) [inline tests/TranslatorTest.php:602]: Why result of "with formatter" contain formatted text?

Comment (vjik) [reply tests/TranslatorTest.php:602]: > Why result of "with formatter" contain formatted text?

Because in category used `NullMessageFormatter`.

Comment (vjik) [reply tests/TranslatorTest.php]: Improved test

**Review by darkdef — APPROVED** [review]

File: src/Translator.php [hunk 2]
```diff
@@ -141,8 +142,12 @@ private function translateUsingCategorySources(
             }
         }
 
-        $categorySource = end($this->categorySources[$category]);
-        return $categorySource->format($id, $parameters, $locale);
+        return end($this->categorySources[$category])->format(
```
Comment (arogachev) [inline src/Translator.php:145]: (Optional, code style) Separate variable for `$categorySource` is better in my opinion in terms of code readability.

File: tests/CategoryTest.php [hunk 3]
```diff
@@ -32,6 +34,40 @@ public function testNameException(): void
...
+                new SimpleMessageFormatter(),
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider dataWithoutFormatter
+     */
+    public function testWithoutFormatter(
+        string $expectedMessage,
```
Comment (arogachev) [inline tests/CategoryTest.php:59]: (Optional) Maybe swap places for the first two arguments? It's more natural for me to see "before" state first, then "after" state.

File: tests/SimpleMessageFormatterTest.php [hunk 4]
```diff
@@ -112,6 +112,15 @@ public function testFormatPluralWithMissingKeys(): void
         $formatter->format('{min, plural}', ['min' => 1]);
     }
 
+    public function testFormatPluralWithMissingKeys2(): void
```
Comment (arogachev) [inline tests/SimpleMessageFormatterTest.php:115]: (Naming) Maybe give more meaningful name?

Comment (arogachev) [reply tests/SimpleMessageFormatterTest.php:115]: By the way, it's completely the same as the test without "2" digit. Maybe refactor it? It does not depend on the order. just the count, right?

File: tests/TranslatorTest.php [hunk 5]
```diff
@@ -559,6 +559,63 @@ public function testTranslationMissingMessage(): void
         $translator->translate('missing_message', [], 'app');
     }
 
+    public function testDefaultMessageFormatterWithCategory(): void
+    {
+        $categorySourceWithFormatter = new CategorySource(
```
Comment (arogachev) [inline tests/TranslatorTest.php:564]: Maybe use data provider?

**Review by arogachev — CHANGES_REQUESTED** [review]
:+1:

File: src/Translator.php (hunk 2 above)
Comment (vjik) [reply src/Translator.php:145]: `end($this->categorySources[$category])` — get last category source. it's good readability for me. This don't use later, I think use variable is overhead.

File: tests/CategoryTest.php (hunk 3 above)
Comment (vjik) [reply tests/CategoryTest.php:59]: In phpunit use before expected value, then actual value: `assertSame(mixed $expected, mixed $actual)`.

File: src/Translator.php (hunk 2 above)
Comment (arogachev) [reply src/Translator.php:145]: Sorry, don't agree. But I marked this as optional.

File: tests/CategoryTest.php (hunk 3 above)
Comment (arogachev) [reply tests/CategoryTest.php:59]: Yes, also thought about that, but this is a data provider's data, not assert. Again, it's optional and up to you.

File: tests/TranslatorTest.php (hunk 5 above)
Comment (vjik) [reply tests/TranslatorTest.php]: Done.

File: tests/SimpleMessageFormatterTest.php (hunk 4 above)
Comment (vjik) [reply tests/SimpleMessageFormatterTest.php]: Refactor to data provider.

**Review by arogachev — APPROVED** [review]

**Comment by arogachev** [comment]
Is BC break expected?

**Comment by arogachev** [comment]
> Is BC break expected?

Discussed in group chat - it's OK.

**Pull request merged**